### PR TITLE
feat(revisions): adopt aep-162 revisions

### LINKED ISSUE #16: Adopt AIP-0162 Resource Revisions

### PR DESCRIPTION
**Pull request by toumorokoshi**
Adapted from aip.dev/162.

list of changes include:

- changing references to "name" to "path".
- minor formatting changes like consolidating proto examples so we can minimize the proto/openapi tabs.
- removed some sections from rationale that spoke to aip.dev more than aeps, and didn't provide valuable context (like the existence of "tag" in older versions of the pattern).
- added openapi schemas.

fixes #16.

## 🍱 Types of changes

What types of changes does your code introduce to AEP? _Put an `x` in the boxes
that apply_

- [ ] Enhancement
- [ ] [New proposal](https://aep.dev/1#workflow)
- [x] Migrated from google.aip.dev
- [ ] Chore / Quick Fix

## 📋 Your checklist for this pull request

Please review the [AEP Style and Guidance](https://aep.dev/style-guide) for
contributing to this repository.

### General

- [x] Basic [Guidance](https://aep.dev/style-guide#guidance) is met.
- [x] Ensure that your PR
      [references AEPs](https://aep.dev/style-guide#referencing-aeps)
      correctly.
- [x] [My code has been formatted](https://aep.dev/contributing#formatting)
      (usually `prettier -w .`)

### Additional checklist for a new AEP

- [x] A new AEP **should** be no more than two pages if printed out.
- [x] Ensure that the PR is editable by maintainers.
- [x] Ensure that [File structure](https://aep.dev/style-guide#file-structure)
      guidelines are met.
- [x] Ensure that
      [Document structure](https://aep.dev/style-guide#document-structure)
      guidelines are met.

💝 Thank you!

### CODE REVIEW

File: aep/general/0162/aep.md.j2 [hunk 1]
```diff
@@ -1,5 +1,396 @@
...
+  contain a list of resource IDs that the revision is also known by (e.g. `latest`)
+
+### Creating Revisions
+
+Depending on the resource, different APIs may have different strategies for
+creating a new revision. Specifically examples of strategies include:
+
+- Creating a revision when there is a change to the parent resource
+- Creating a revision when important system state changes
+- Creating a revision when specifically requested
```
Comment (toumorokoshi) [inline aep/general/0162/aep.md.j2:107]: one things to note is that I do believe that this provision allows one to create a revision object with a create method, with a user-settable id. The value would just be the id, with the snapshot being whatever the resource was at the time the revision was created.

This does allow the resource to be managed via Declarative clients, although I'm not 100% sure what the value would be there (perhaps if constructing a resource requires that it references a revision and not the actual resource).

File: aep/general/0162/aep.md.j2 [hunk 2]
```diff
@@ -1,5 +1,396 @@
...
+### Resource names for revisions
+
+When referring to specific revision of a resource, the subcollection name
+**must** be named `revisions`. Resource revisions have paths with the format
+`{resource_name}/revisions/{revision_id}`. For example:
+```
+publishers/123/books/les-miserables/revisions/c7cfa2a8
+```
+
+### Server-specified Aliases
```
Comment (toumorokoshi) [inline aep/general/0162/aep.md.j2:121]: since revisions does allow for user and server specified ids to live side by side with each other - I think we probably want a new AEP on that specifically.

I already have some thoughts on this - we can propose a namespacing scheme as a *should*, and provide that as a way of supporting both user-specified and non-user-specified resources.

File: aep/general/0162/aep.md.j2 [hunk 3]
```diff
@@ -1,5 +1,396 @@
...
+resource by using `@` symbol. List and delete revisions were custom methods on
+the resource collection. A single Get method was used to retrieve either the
+resource revision, or the resource.
+
+Its primary advantage was allowing a resource reference to seamlessly refer to
+a resource, or its revision.
+
+It also had several disadvantages:
+
+-  List revisions is a custom method (:listRevisions) on the resource collection
```
Comment (rofrankel) [inline aep/general/0162/aep.md.j2:370]:
```suggestion
-  List revisions is a custom method (`:listRevisions`) on the resource.
```

Why would it be on a collection?

File: aep/general/0162/aep.md.j2 [hunk 4]
```diff
@@ -1,5 +1,396 @@
...
+-  Delete revision is a custom method on the resource collection
```
Comment (rofrankel) [inline aep/general/0162/aep.md.j2:371]:
```suggestion
-  Delete revision is a custom method on the resource.
```

Why would it be on a collection?

File: aep/general/0162/aep.md.j2 [hunk 5]
```diff
@@ -1,5 +1,396 @@
...
+-  Not visible in API discovery doc
```
Comment (rofrankel) [inline aep/general/0162/aep.md.j2:372]:
```suggestion
-  Not visible in API discovery documenation.
```

File: aep/general/0162/aep.md.j2 [hunk 6]
```diff
@@ -1,5 +1,396 @@
...
+-  Resource ID cannot use `@`
```
Comment (rofrankel) [inline aep/general/0162/aep.md.j2:373]:
```suggestion
-  Resource IDs cannot use the `@` symbol.
```

FWIW, we worked around this at Roblox as follows:

* Only the last `@` is assumed to specify a rev ID.
* If you have a `@` in your resource ID and don't want a revision then you have to add `@latest` to the end.  (But this wouldn't work  if we had a different strategy for when to create revisions.)

File: aep/general/0162/aep.md.j2 [hunk 7]
```diff
@@ -1,5 +1,396 @@
...
+
+The guidance was modified ultimately to enable revisions to behave like a
+resource, which reduces the users cognitive load and allows resource-oriented
```
Comment (rofrankel) [inline aep/general/0162/aep.md.j2:376]:
```suggestion
resource, which reduces users' cognitive load and allows resource-oriented
```

File: aep/general/0162/aep.md.j2 [hunk 8]
```diff
@@ -1,5 +1,396 @@
...
+clients to easily list, get, create, and update revisions.
```
Comment (rofrankel) [inline aep/general/0162/aep.md.j2:377]:
```suggestion
clients to easily list, get, create, update, and delete revisions.
```

**Review by rofrankel — CHANGES_REQUESTED** [review]

File: aep/general/0162/aep.md.j2 [hunk 9]
```diff
@@ -1,5 +1,396 @@
...
+reason about the state of the resource over time. There are several reasons for
+this:
+
+- Users may want to be able to roll back to a previous revision, or diff
+  against a previous revision.
+- An API may create data which is derived in some way from a resource at a
+  given point in time. In these cases, it may be desirable to snapshot the
+  resource for reference later.
+
+**Note:** We use the word _revision_ to refer to a historical reference for a
```
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0162/aep.md.j2:13]: I think the issue with "version" is that someone could be referring to V2 of the resource schema, vs V2 of the resource itself. 

I personally don't think there's a cut-and-dry term here - version *could* be ok, but will definitely lead to several conversations where someone has to clarify which concept of "version" they're referring to.

[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0162/aep.md.j2:13]: If you feel strongly against "revision" and feel strongly for "snapshots", I'm fine with snapshots.

File: aep/general/0162/aep.md.j2 [hunk 10]
```diff
@@ -1,5 +1,396 @@
...
+useful include:
+
+- When it is valuable to expose older revisions of a resource via an API. This
+  can avoid the overhead of the customers having to write their own API to store
+  and enable retrieval of revisions.
+- Other resources depend on different revisions of a resource.
+- There is a need to represent the change of a resource over time.
+
+APIs implementing resources with a revision **should** abstract resource
+revisions as nested collection of the resource. Sometimes, the revisions
```
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0162/aep.md.j2:29]: either is fine by me!

File: aep/general/0162/aep.md.j2 [hunk 11]
```diff
@@ -1,5 +1,396 @@
...
+- The resource revision **must** contain a `create_time` field (see [AIP-142][]).
+- The resource revision **may** contain a repeated field `alternate_ids`, which would
+  contain a list of resource IDs that the revision is also known by (e.g. `latest`)
+
+### Creating Revisions
+
+Depending on the resource, different APIs may have different strategies for
+creating a new revision. Specifically examples of strategies include:
+
+- Creating a revision when there is a change to the parent resource
```
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0162/aep.md.j2:105]: FWIW I'm not sure how often the carveout is used. If you want me to remove it - I'm ok with that for now (and we can always re-add when there's a valid use case).

File: aep/general/0162/aep.md.j2 [hunk 12]
```diff
@@ -1,5 +1,396 @@
...
+- Creating a revision when specifically requested
+
+APIs **may** use any of these strategies. APIs **must**
+document their revision creation strategy.
+
+### Resource names for revisions
+
+When referring to specific revision of a resource, the subcollection name
+**must** be named `revisions`. Resource revisions have paths with the format
+`{resource_name}/revisions/{revision_id}`. For example:
```
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0162/aep.md.j2:116]: yes - it should match the path syntax. even if it's not explicitly AEP path syntax, there's no good reason to introduce a new one.

I don't disagree with this matching the syntax - but I do wonder if we should revisit the whole don't-suffix-with-id thing (it's an issue that came up with the Terraform provider). I'll leave it open so we can talk more about it offline.

File: aep/general/0162/aep.md.j2 [hunk 13]
```diff
@@ -1,5 +1,396 @@
...
+  - The field **must** be [annotated as required][aip-203].
+- If the user calls the method with an existing `alias_id`, the request **must**
+  succeed and the alias will be updated to refer to the provided revision. This
+  allows users to write code against a specific alias (e.g. `published`) and the
+  revision can change with no code change.
+
+### Rollback
+
+A common use case for a resource with a revision history is the ability to roll
+back to a given revision. APIs **should** handle this with a `Rollback` custom
```
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0162/aep.md.j2:229]: why a may? I think it's saying "if you're going to support rollbacks, do it this way". 

That said most other AIPs use *may* here, so I'm not exactly against it - but maybe something we want to clarify in a voice guide?

File: aep/general/0162/aep.md.j2 [hunk 14]
```diff
@@ -1,5 +1,396 @@
...
+APIs **should not** include multiple levels of resources with revisions, as
+this quickly becomes difficult to reason about.
+
+### Standard methods
+
+Any standard methods **must** implement the corresponding AIPs (AIP-131,
+AIP-132, AIP-133, AIP-134, AIP-135), with the following additional behaviors:
+
+- List methods: By default, revisions in the list response **must** be ordered
+  in reverse chronological order. User can supply `order_by` to override the
```
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0162/aep.md.j2:328]: it doesn't not.... but I guess that's an AEP generator bug we should just fix.

[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0162/aep.md.j2:328]: https://github.com/aep-dev/site-generator/issues/28

File: aep/general/0162/aep.md.j2 (hunk 6 above)
Comment (toumorokoshi) [reply aep/general/0162/aep.md.j2:373]: Sure, perhaps it's not the main reason. It's arguable that revisions are a carve-out and therefore able to not follow the AEPs in this regard - but you don't really want a lot of carve-outs in your spec.

File: aep/general/0162/aep.md.j2 [hunk 15]
```diff
@@ -1,5 +1,396 @@
...
+
+APIs **should not** include multiple levels of resources with revisions, as
+this quickly becomes difficult to reason about.
+
+### Standard methods
+
+Any standard methods **must** implement the corresponding AIPs (AIP-131,
+AIP-132, AIP-133, AIP-134, AIP-135), with the following additional behaviors:
+
+- List methods: By default, revisions in the list response **must** be ordered
```
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0162/aep.md.j2:327]: FWIW I don't like this anyway - I think the order of list should be consistent. I'd prefer to remove this exceptional behavior, but I do see the point that you may want to get it newest-first.

That said there's always a query parameter that can be added to sort as you wish.

File: aep/general/0162/aep.md.j2 [hunk 16]
```diff
@@ -1,5 +1,396 @@
...
+
+- When it is valuable to expose older revisions of a resource via an API. This
+  can avoid the overhead of the customers having to write their own API to store
+  and enable retrieval of revisions.
+- Other resources depend on different revisions of a resource.
+- There is a need to represent the change of a resource over time.
+
+APIs implementing resources with a revision **should** abstract resource
+revisions as nested collection of the resource. Sometimes, the revisions
+collection can be a top level collection. Exceptions include:
```
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0162/aep.md.j2:30]: yeah you're right - I think "sibling to the resource" is probably what was intended here. will change.

[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0162/aep.md.j2:30]: I tend to agree. This isn't a snapshot exactly - but one example of a sibling revision could be databases. You don't want to *delete* a database - you might want to restore it.

But thinking further - even in the database case, I think you would just have a custom method "restore" with the old ID. So there's probably other ways to address the need for revisions that outlive their resource.

The way I look at these, we can remove examples until we have a good use case, and always re-add an improved one with the wisdom on an actual need. So - should I just remove all references to non-child revisions?

File: aep/general/0162/aep.md.j2 [hunk 17]
```diff
@@ -1,5 +1,396 @@
...
+
+{% tab proto %}
+
+```proto
+message BookRevision {
+  // The path of the book revision.
+  string path = 1;
+
+  // The snapshot of the book
+  Book snapshot = 2
```
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0162/aep.md.j2:43]: I'd be fine with `book`. But personally I hate that we landed on the resource singular for the resource, instead of a standard name.

Even for the CRUD methods I'd prefer something more generic, like "resource", rather than the resource singular - it vastly simplifies automation which would have to go extract the resource singular to know what the right field to extract the resource would be.

File: aep/general/0162/aep.md.j2 (hunk 1 above)
Comment (toumorokoshi) [reply aep/general/0162/aep.md.j2:107]: there's no clause that says you *can't* do that - and actually it's a good missing example! I'll add it as one of the possible times to create a revision.

Do we need to go as far as requiring at least one revision is created? doesn't seem strictly necessary, but I imagine most cases for resource revisions would benefit from one always being created and can therefore be referenced.

File: aep/general/0162/aep.md.j2 [hunk 18]
```diff
@@ -1,5 +1,396 @@
...
+- The request message **must** have a `path` field, referring to the resource
+  revision whose configuration the resource should be rolled back to.
+  - The field **must** be [annotated as required][aip-203].
+  - The field **must** identify the [resource type][aip-123] that it
+    references.
+
+### Child resources
+
+Resources with a revision history **may** have child resources. If they do,
+there are two potential variants:
```
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0162/aep.md.j2:312]: no - I don't think there's good guidance here. At least nothing that is generic - it'd probably be very case-dependent, and the API author would make a better judgement probably.

That said - it's good to enumerate your options here - some may not realize that a revision may also snapshot children.

File: aep/general/0162/aep.md.j2 [hunk 19]
```diff
@@ -1,5 +1,396 @@
...
+
+- Child resources where each child resource is a child of the parent resource
+  as a whole.
+- Child resources where each child resource is a child of _a single revision
+  of_ the parent resource.
+
+APIs **should not** include multiple levels of resources with revisions, as
+this quickly becomes difficult to reason about.
```
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0162/aep.md.j2:320]: I'm fine with removing it - I agree there isn't a good enough rule here to be worth putting this guidance in.

File: aep/general/0162/aep.md.j2 [hunk 20]
```diff
@@ -1,5 +1,396 @@
...
+
+## Rationale
+
+### Abstract revisions as nested collection
+
+Revisions being resources under nested collection make revisions a first class
+citizen.
+
+- Revisions can offer standard get, list, and delete methods.
+- It retains the flexibility of extending new fields to a revision in addition to
```
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0162/aep.md.j2:344]: I can just rephrase it - "retains" isn't useful context.

I'm sure it's just from the fact that we went from v1 -> v2.

File: aep/general/0162/aep.md.j2 [hunk 21]
```diff
@@ -1,5 +1,396 @@
...
+  the resource message.
+
+### Output only resource configuration
+
+Although it was an option to have the revision take in the resource
+configuration as part of the create method, doing so would have allowed users to
+submit resource configuration for a revision that the resource was never in.
```
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0162/aep.md.j2:352]: basically - the snapshot field is read-only because if it was mutable, someone could create a snapshot with values that the resource never had - a fake snapshot, basically.

File: aep/general/0162/aep.md.j2 (hunk 1 above)
Comment (rofrankel) [reply aep/general/0162/aep.md.j2:107]: Yeah, I think it's a bit weird to have a resource with no revisions - kind of like a git repo with no commits.  I'm open to being persuaded otherwise but my current inclination is that the list of revisions should never be empty.

File: aep/general/0162/aep.md.j2 (hunk 9 above)
[the remark this answers is not in the document]
Comment (rofrankel) [reply aep/general/0162/aep.md.j2:13]: > I think the issue with "version" is that someone could be referring to V2 of the resource schema, vs V2 of the resource itself.

I know this is the rationale, but I don't buy that this is something anyone's likely to get confused about in practice.  Maybe I'm just failing to think of a scenario/conversation where there's actual ambiguity?

Re "snapshots"...are there cases where "snapshot" might be used for another pattern, and so it would create the kind of ambiguity that is allegedly a risk with "version"?  I have a specific example in mind (but not sure I can put it in writing somewhere public at the moment).

[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0162/aep.md.j2:13]: > I know this is the rationale, but I don't buy that this is something anyone's likely to get confused about in practice. Maybe I'm just failing to think of a scenario/conversation where there's actual ambiguity?

For me it's the fact that there'd be conversations around a resource with revisions that may have a v1 / v2 API, and then it gets really confusing (e.g. "version 2 of the item version api" would be a valid sentence).

> Re "snapshots"...are there cases where "snapshot" might be used for another pattern, and so it would create the kind of ambiguity that is allegedly a risk with "version"? I have a specific example in mind (but not sure I can put it in writing somewhere public at the moment).

As is the case with most English words, there's plenty of opportunities for overloaded language and conflicts. FWIW I don't think we have to think much deeper beyond this - We should just decide on nomenclature that's consistent with the AEPs.

One reason to leave it as "revisions" is that aip.dev already defines it that way, and our most likely onboarded to AEPs are those that are using the AIPs already.

File: aep/general/0162/aep.md.j2 (hunk 1 above)
Comment (toumorokoshi) [reply aep/general/0162/aep.md.j2:107]: If a company comes along and they want to adopt the AEPs, and they have a resource with no revisions, would we block their adoption of AEPs based on this reason? 

I'd argue no: although I see a minor point on usability, I don't see a reason why this makes a client less usable.

I'd be open to make this a should though?

File: aep/general/0162/aep.md.j2 (hunk 20 above)
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0162/aep.md.j2:344]: just going to remove this paragraph - it doesn't make sense without considering alternatives or history.

File: aep/general/0162/aep.md.j2 (hunk 17 above)
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0162/aep.md.j2:43]: I filed a larger issue on this - I think I'd like to argue we go forward with standardizing and always calling it "snapshot": https://github.com/aep-dev/aep.dev/issues/205

File: aep/general/0162/aep.md.j2 (hunk 1 above)
Comment (rofrankel) [reply aep/general/0162/aep.md.j2:107]: I think it would be suboptimal if users of multiple AEP-compliant APIs from different organizations had to learn this kind of difference between the two.  I agree it probably doesn't break generated clients.

Maybe a **should** is the right balance.

File: aep/general/0162/aep.md.j2 (hunk 9 above)
[the remark this answers is not in the document]
Comment (rofrankel) [reply aep/general/0162/aep.md.j2:13]: >  "version 2 of the item version api" would be a valid sentence

Would it?  The API isn't an "item version" API, it's an "item" API where `Item` supports versions.  So "item version API" seems to me like an incorrect name.

I don't think "revision" is terrible, but I think approximately everyone at Roblox has either gotten confused by it, or asked why it's not "version", or just says "version" instead of "revision".  And I don't really want to defend it with "APIs can also be versioned" because I don't find it persuasive.

**Review by rofrankel — COMMENTED** [review]
Just taking myself out of the attention set

File: aep/general/0162/aep.md.j2 (hunk 17 above)
[the remark this answers is not in the document]
Comment (rofrankel) [reply aep/general/0162/aep.md.j2:43]: I guess there are two topics here:

1. "<resource>" vs. something static
2. Which static thing?

I'm inclined to agree on (1), even though the implications are inconvenient for me/Roblox (extrapolating that we'll conclude the same for e.g. CreateBookRequest).

For (2) though:  why "snapshot" and not "revision" or "resource"?

File: aep/general/0162/aep.md.j2 (hunk 13 above)
[the remark this answers is not in the document]
Comment (rofrankel) [reply aep/general/0162/aep.md.j2:229]: Ah, I interpreted it differently:  "APIs **should** support this, and here's how...".

New suggestion, then:

```suggestion
back to a given revision. APIs that support this behavior **should** do so with a `Rollback` custom
```

File: aep/general/0162/aep.md.j2 (hunk 18 above)
[the remark this answers is not in the document]
Comment (rofrankel) [reply aep/general/0162/aep.md.j2:312]: Some API owners or some API clients?

Maybe the guidance is:  if a revision resource like BookRevision has descendants, they must be a subset of the descendants of Book, and a given BookRevision's descendants must be the same as the descendants of the Book at the time the revision was created (or perhaps some limited view of those resources).

It doesn't have to be exactly that, but I do prefer something that that which is actual guidance.

File: aep/general/0162/aep.md.j2 (hunk 21 above)
[the remark this answers is not in the document]
Comment (rofrankel) [reply aep/general/0162/aep.md.j2:339]: Ah, I understand now.  Some suggestions that would help make that clearer:

* Fix the hyphen placement in the section header ("Output-only resource..." rather than "Output only-resource...")
* Remove replace "resource configuration" (which is not a defined term AFAIK) with something else, maybe "snapshots"
* Break it up into two sentences and kill the trailing preposition (I'm not puritanical about trailing prepositions, but in this case I think it does contribute to the sentence being hard to parse).

That said...do we need to explain this guidance in particular?  It seems pretty obvious to me why the snapshot field would be output-only.

File: aep/general/0162/aep.md.j2 (hunk 1 above)
Comment (toumorokoshi) [reply aep/general/0162/aep.md.j2:107]: SGTM, added.

File: aep/general/0162/aep.md.j2 (hunk 9 above)
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0162/aep.md.j2:13]: FWIW I'd probably argue that it's really up to someone to argue why an existing precedent *isn't* ok, rather than justify that precedent. Why isn't revision ok? In my experience at Google, I never heard an issue with that naming choice.

Maybe @wora might have an opinion on this as well as a third opinion?

File: aep/general/0162/aep.md.j2 (hunk 17 above)
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0162/aep.md.j2:43]: sounds good, focusing on 2 (1 we can discuss in #205): My vote would be for "resource" - then it could even use the same noun as the standard methods, which would be amazingly consistent.

File: aep/general/0162/aep.md.j2 (hunk 18 above)
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0162/aep.md.j2:312]: okay sounds good. I like your guidance - let's add that. I'll remove the "enumeration of options" part of things too.

File: aep/general/0162/aep.md.j2 (hunk 21 above)
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0162/aep.md.j2:339]: I'm fine with removing it!

File: aep/general/0162/aep.md.j2 [hunk 22]
```diff
@@ -1,5 +1,372 @@
...
+
+{% endtabs %}
+
+- The request message **must** have a `path` field:
+  - The field **must** be [annotated as required](/field-behavior-documentation).
+  - The field **must** identify the [resource type](/resource-types) that it
+    references.
+- The request message **must** have a `alias` field:
+  - The field **must** be [annotated as required][aip-203].
+- If the user calls the method with an existing `alias`, the request **must**
```
Comment (dv-stewarts) [inline aep/general/0162/aep.md.j2:223]: Some tools allow this kind of tagging as described here. Git, notably, errors if you tag a different commit with an existing tag, requiring a `-f` flag to override. Should this behaviors be supported?

This could be introduced as a non-breaking addition by including in the specification an option like `--force-with-lease` which errors if the alias already exists.

**Review by dv-stewarts — COMMENTED** [review]
Mixture of thoughts and typo/terminology nitpicks.

File: aep/general/0162/aep.md.j2 [hunk 23]
```diff
@@ -1,5 +1,372 @@
...
+{% endtabs %}
+
+- The method **must** use the `POST` HTTP verb.
+- The method **should** return a resource revision.
+- The request message **must** have a `path` field, referring to the resource
+  revision whose configuration the resource should be rolled back to.
+  - The field **must** be [annotated as required][aip-203].
+  - The field **must** identify the [resource type][aip-123] that it
+    references.
+
```
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0162/aep.md.j2:310]: Talking to @rofrankel, I think we've decided to be more concise where possible - so reading the AEPs, assume anything not explicitly stated are not dictated.

File: aep/general/0162/aep.md.j2 [hunk 24]
```diff
@@ -1,5 +1,372 @@
...
+### Child resources
```
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0162/aep.md.j2:311]: maybe we can split this up? I think there's two points here: not understanding the existing guidance, and a question about a specific scenario that isn't explicit enough in the guidance.

For the former: what part is unclear? Is there a specific sentence we can make clearer, or a sentence that may have more than one meaning?

For the latter: we could (and probably should) be explicit about this, but if you're saying:

1. user creates resource Bookshelf, with subresource Book. (/bookshelf/x/book/y)
2. user creates a revision "A" of the Bookshelf, which contains the Book. (/bookshelf/x/revision/a/). The revision also stores it's subresources (/bookshelf/x/revision/a/book/y)
3. user deletes Book.

I'd hope that when user does a get on (/bookshelf/x/revision/a/book/y), the get the book back (since it's a subresource of the revision).

WDYT? We could definitely add more precise guidance on this case - though I'd like to do that as a follow-up PR.

File: aep/general/0162/aep.md.j2 [hunk 25]
```diff
@@ -1,5 +1,372 @@
...
+```
+
+{% endtabs %}
+
+- The resource revision **must** contain a field with a message type of the
+  parent resource, with a field name of `resource`.
+    - The value of `resource` **must** be the original resource
+      at the point in time the revision was created.
+- The resource revision **must** contain a `create_time` field (see [AIP-142][]).
+- The resource revision **may** contain a repeated field `aliases`, which would
```
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0162/aep.md.j2:97]: Sorry about that - this was a type, and "aliases" needs to be updated in the openapi. I've tried to correct all the references I can.

The idea with aliases is to align it with the existing concept in the A(IE)Ps: https://aep.dev/122#resource-id-aliases

I think regarding the tag - you're right, we are unfortunatleye repeating ourselves. It is funny - I wrote that and I think I wrote that when we didn't re-introduce "aliases" because of some internal Google feedback.

I guess the best option here is to just delete the section about tags in the history - at least now we're using a consistent term with AEPs for this (alias).

File: aep/general/0162/aep.md.j2 (hunk 22 above)
Comment (toumorokoshi) [reply aep/general/0162/aep.md.j2:223]: Good call! I like that. Added a field "overwrite". I think "force" isn't very explicit w.r.t. what behavior it's overwriting. thoughts?

File: aep/general/0162/aep.md.j2 (hunk 2 above)
Comment (toumorokoshi) [reply aep/general/0162/aep.md.j2]: adding a *should* about that here.

File: aep/general/0162/aep.md.j2 (hunk 9 above)
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0162/aep.md.j2:13]: after a few offline discussions in the weekly meeting, we've agreed to "revision" - the original terminology.

File: aep/general/0162/aep.md.j2 (hunk 25 above)
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0162/aep.md.j2:97]: removed section about tags.

File: aep/general/0162/aep.md.j2 [hunk 26]
```diff
@@ -1,5 +1,372 @@
...
+### Creating Revisions
+
+Depending on the resource, different APIs may have different strategies for
+creating a new revision. Specifically examples of strategies include:
+
+- Creating a revision when there is a change to the resource
+- Creating a revision when important system state changes
+- Creating a revision when specifically requested
+
+APIs **may** use any of these strategies. APIs **must**
```
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0162/aep.md.j2]: applied! thanks.

File: aep/general/0162/aep.md.j2 [hunk 27]
```diff
@@ -1,5 +1,372 @@
...
+### User-specified aliases
+
+APIs **may** provide a mechanism for users to assign an [alias][] ID to an
+existing revision with a custom method "alias":
+
+{% tab proto %}
+
+```proto
+
+rpc AliasBookRevision(TagBookRevisionRequest) returns (Book) {
```
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0162/aep.md.j2]: thanks! applied.

File: aep/general/0162/aep.md.j2 (hunk 22 above)
Comment (toumorokoshi) [reply aep/general/0162/aep.md.j2]: resolving for now, but feel free to re-open if you have concerns about the name.

File: aep/general/0162/aep.md.j2 (hunk 16 above)
[the remark this answers is not in the document]
Comment (toumorokoshi) [reply aep/general/0162/aep.md.j2]: removed it for now.

File: aep/general/0162/aep.md.j2 [hunk 28]
```diff
@@ -106,9 +102,11 @@ creating a new revision. Specifically examples of strategies include:
 - Creating a revision when important system state changes
 - Creating a revision when specifically requested
 
-APIs **may** use any of these strategies. APIs **must**
+APIs **may** use any of these strategies or any other strategy. APIs **must**
 document their revision creation strategy.
 
+APIs **should** have at least one revision for their resource at any time.
```
Comment (rofrankel) [inline aep/general/0162/aep.md.j2:108]:
```suggestion
Resources that support revisions **should** always have at least one revision.
```

File: aep/general/0162/aep.md.j2 [hunk 29]
```diff
@@ -127,6 +125,8 @@ Services **may** reserve specific IDs to be [aliases][alias] (e.g.
 GET /v1/publishers/{publisher_id}/books/{book_id}/revisions/{revision_id}
 ```
 
+If specific IDs are reserved, services **should** document those IDs.
```
Comment (rofrankel) [inline aep/general/0162/aep.md.j2:128]: Why not must?

File: aep/general/0162/aep.md.j2 [hunk 30]
```diff
@@ -332,14 +336,25 @@ As revisions are nested under the resource, also see [cascading delete][].
 
 ## Rationale
 
-### Output only-resource configuration
+### For the name revision
```
Comment (rofrankel) [inline aep/general/0162/aep.md.j2:339]:
```suggestion
### For the name "revision"
```

File: aep/general/0162/aep.md.j2 [hunk 31]
```diff
@@ -1,5 +1,388 @@
...
+
+APIs **may** use any of these strategies or any other strategy. APIs **must**
+document their revision creation strategy.
+
+APIs **should** have at least one revision for their resource at any time.
+
+### Resource names for revisions
+
+When referring to specific revision of a resource, the subcollection name
+**must** be named `revisions`. Resource revisions have paths with the format
```
Comment (rofrankel) [inline aep/general/0162/aep.md.j2:113]:
```suggestion
**must** be `revisions`. Resource revisions have paths with the format
```

File: aep/general/0162/aep.md.j2 [hunk 32]
```diff
@@ -1,5 +1,388 @@
...
+- Creating a revision when specifically requested
+
+APIs **may** use any of these strategies or any other strategy. APIs **must**
+document their revision creation strategy.
+
+APIs **should** have at least one revision for their resource at any time.
+
+### Resource names for revisions
+
+When referring to specific revision of a resource, the subcollection name
```
Comment (rofrankel) [inline aep/general/0162/aep.md.j2:112]:
```suggestion
When referring to the revisions of a resource, the subcollection name
```

Comment (rofrankel) [reply aep/general/0162/aep.md.j2:112]: Do we also want something here about how the resource name of the revision resource must be "{Resource}Revision", i.e. "BookRevision"?

Otherwise this could also be interpreted as saying that the resource has to be named "Revision".  This obviously doesn't work for APIs with multiple revisioned resources...but doesn't hurt to spell it out.

**Review by rofrankel — CHANGES_REQUESTED** [review]

File: aep/general/0162/aep.md.j2 (hunk 32 above)
Comment (toumorokoshi) [reply aep/general/0162/aep.md.j2:112]: Are there exceptional cases to this? I guess not if the path includes `revision` in the name. Added.

File: aep/general/0162/aep.md.j2 (hunk 29 above)
Comment (toumorokoshi) [reply aep/general/0162/aep.md.j2:128]: sure, isn't really an enforceable must though.

Comment (toumorokoshi) [reply aep/general/0162/aep.md.j2]: added

File: aep/general/0162/aep.md.j2 [hunk 33]
```diff
@@ -1,5 +1,390 @@
...
+### Resource names for revisions
+
+When referring to the revisions of a resource, the subcollection name
+**must** be `revisions`. Resource revisions have paths with the format
+`{resource_path}/revisions/{resource_revision}`. For example:
+```
+publishers/123/books/les-miserables/revisions/c7cfa2a8
+```
+
+The resource **must** be named `{resource_singular}Revision`.
```
Comment (rofrankel) [inline aep/general/0162/aep.md.j2:119]:
```suggestion
The resource **must** be named `{resource_singular}Revision` (for example, `BookRevision`).
```

**Review by rofrankel — APPROVED** [review]
